risc0/bigint2/src/rsa/mod.rs

```diff
@@ -62,8 +62,8 @@ pub fn modpow_65537(base: &BigUint, modulus: &BigUint) -> BigUint {
 }
 
 pub fn raw_modpow_65537(base: &[u32], modulus: &[u32], result: &mut [u32]) {
-    assert_eq!(base.len(), RSA_3072_WIDTH_WORDS);
-    assert_eq!(modulus.len(), RSA_3072_WIDTH_WORDS);
+    assert!(base.len() <= RSA_3072_WIDTH_WORDS);
+    assert!(modulus.len() <= RSA_3072_WIDTH_WORDS);
```

Comment:
I don't think it's safe to call this function with lengths < RSA_3072_WIDTH_WORDS, and I'm surprised `modpow_65537_small_buffers` doesn't break. What this is telling the accelerator to do is read 3072 bits starting from the address `base.as_ptr()` and read 3072 bits starting from the address `modulus.as_ptr()`, do modpow_65537 on them, and write the resulting 3072 bits to `result.as_mut_ptr()`. I really don't see how that could be safe if any of `base`, `modulus`, or `result` aren't the full 3072 bits. Maybe the test is getting "lucky" and has a bunch of zeros after each of `base` and `modulus`?

Oh, maybe we're getting the right answer because the prime `7` is so small, so it's relatively easy to get an answer by shear luck. How about trying out a roughly 32 bit prime instead of 7? You could use BabyBear, for example. Or maybe we need to have more code in the guest, so that when we read too far we run into something other than zeros.

Reply:
(I don't think you need to know this to make these changes, but if you want context on how much will be read: Look at https://github.com/risc0/zirgen/blob/main/zirgen/circuit/bigint/bigint2c.cpp#L405 and the surrounding lines. `11`, `12`, and `13`, used in the `LoadOp`s and `StoreOp` indicate reading/writing to the 1st, 2nd, and 3rd buffers provided in `sys_bigint2_3`, and the `bits = 3072` used with the loads controls how much will be read.)

Reply:
Agreed that these asserts should not be changed, they need to check that the full 3072 bits are available to be read.

Reply:
Trying to debug the fact that the tests pass, I got a result that was unexpected. Digging in more to understand, as I might just be misunderstanding something.

Printing the memory that's used for the syscall without fixing the buffer sizes:

```
    let base = to_u32_digits(base);
    let modulus = to_u32_digits(modulus);
    let _tmp = std::hint::black_box(Box::new([0u32; RSA_3072_WIDTH_WORDS]));
    unsafe {
        let modulus_ptr = modulus.as_ptr();
        let base_ptr = base.as_ptr();

        println!("Memory after modulus:");
        for i in 0..RSA_3072_WIDTH_WORDS {
            let addr = modulus_ptr.add(i);
            println!("0x{:x}: {:08x}", addr as usize, *addr);
        }

        println!("\nMemory after base:");
        for i in 0..RSA_3072_WIDTH_WORDS {
            let addr = base_ptr.add(i);
            println!("0x{:x}: {:08x}", addr as usize, *addr);
        }
    }

```

Yields:

```
Memory after modulus:
0x230948: 7fffffff
0x23094c: 00000000
0x230950: 00000000
0x230954: 00000000
0x230958: 00000000
0x23095c: 00000000
0x230960: 00000000
0x230964: 00000000
...

Memory after base:
0x230938: 00000001
0x23093c: 00000000
0x230940: 00000000
0x230944: 00000000
0x230948: 7fffffff
0x23094c: 00000000
0x230950: 00000000
...
```

And the test still passes here for some reason. (modulus being `2^31-1`)

In either case, these tests fail if there is data allocated after these are. This test wasn't causing issues previously because there was nothing allocated after these, and memory is zeroed by default. I don't have an explanation for why the above case passes, though

Reply:
It's exponentiation to the 65537 power, not multiplication, and I asked python (which has out-of-the-box modpow):
```
>>> pow(0x7fffffff00000000000000000000000000000001, 65537, 2147483647)
1
```
Finite fields have a lot of structure encoded in their exponentiation (which is how RSA works at all), so you will hit 1 as a result more often than you would just by rolling dice. If instead of the Mersenne prime 2147483647 you used the baby bear prime 2013265921 then python suggests it should make the test fail:
```
>>> pow(0x7fffffff00000000000000000000000000000001, 65537, 2013265921)
606819399
```

Reply:
Discussed in DMs and we realized that my approach doesn't work because the modulus is the thing stomping the memory, so the stomped base value changes. Fundamentally, we're exponentiating (1 + 2^n * modulus)^65537 % modulus which will always be 1. So, what we'll need to do is create a test where the modulus is large enough that the stomped memory overflows the extra memory left in base, thus breaking that formula and causing the test to fail.

Reply:
Resoled in DMs, `pow(0x7fffffff00000000000000000000000000000001, 65537, 2013265921)` turned out as not 1 result because the prime is different in the mod and mangled base, `pow(0x7800000100000000000000000000000000000001, 65537, 2013265921) == 1`

@tzerrell had a great suggestion about using a full length mod value such that only part of the data when read past the buffer is used. Otherwise, can't make `BigUint(1)` large enough to split the value given it's always truncated to 1 word in this implementation. 

Modified test case [`86d0780` (#2510)](https://github.com/risc0/risc0/pull/2510/commits/86d0780fcedf9f0f6a5e98a6adc4ee8cbec62cc4) such that it fails before [`53432e7` (#2510)](https://github.com/risc0/risc0/pull/2510/commits/53432e71cf812097412c0dc5e58e51a18e3a3b26) and also added a test for the inverse pattern, just in case it's ever changed that the mod is allocated before the base